js/util/urlParser.js

```diff
@@ -130,8 +132,64 @@ var urlParser = {
       return encodeURI('file://' + path)
     }
   },
+  getDomain: function (url) {
+    return /^([^:\/]+)/.exec(urlParser.removeProtocol(url))[0].toLowerCase()
+  },
+  validateDomain: function(domain) {
+    // primitive domain validation based on RFC1034
+    domain = punycode.toASCII(domain)
+
+    /*
+      Tests for regex:
+      ^^^^^^^^^^^^^^^
+        google.com              = true
+        google.com.br           = true
+        -google.com             = false
+        google-is-bad.com.ar    = true
+        google.com.             = true
+        google.com-             = false
+        admin:1234@google.com   = true (ignores user:pass)
+        xn--bcher-kva.example   = true
+        127.0.0.1               = true
+        169.254.169.254:8080    = true
+        google.com..            = false
+        [:::1]:8080             = true
+        [2001:0db8::0001:0000]  = true
+        [invalid]               = false
+        localhost               = true
+    */
+    if (!(/^(?!-)(?:.*@)*?([a-z0-9-\._]+[a-z0-9]|\[[\:a-f0-9]+\])(?:\:[0-9]*|\.{1}){0,1}$/i.test(domain) || domain.length > 255)) {
+      return false
+    }
+    let clean_domain = RegExp.$1
+
+    // is domain an ip? then set as valid  - TODO: get a better ipv4-6 regex
+    if (/^((?:[0-9]{1,3}\.){3}[0-9]{1,3}|\[[:a-f0-9]+\])$/i.test(clean_domain))
+      return true
+
+    // known local hostnames
+    if (domain === "localhost" || domain.endsWith('.localhost'))
+      return true
+
+    for (suffix of publicSuffixes) {
+      if (suffix.startsWith('!')) // not supported
+        continue
+      if (suffix.startsWith('*.')) // not supported, simplifying wildcards
+        suffix = suffix.substr(2)
+      if(suffix[0] !== '.')
+        suffix = '.' + suffix
+
+      if (clean_domain.endsWith(suffix))
+        return true
+    }
+
+    return false
+  },
   isHTTPSUpgreadable: function (url) {
-    let domain = /^([^\/]+)/.exec(urlParser.removeProtocol(url))[0]
+    let domain = urlParser.getDomain(url)
+
+    if (urlParser.validateDomain(domain) === false)
+      return false
```

Comment:
If the url isn't a valid domain, it won't be in the top sites list, so this is redundant.